STORM-2870 Properly shutdown ExecutorService in FileBasedEventLogger

HeartSaVioR opened the pull request:
* extract local variable 'scheduler' to one of fields
* gracefully shutdown the scheduler

I followed the guide of Javadoc for implementing `closeFlushScheduler()` properly.
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html

arunmahadevan commented on storm-client/src/jvm/org/apache/storm/metric/FileBasedEventLogger.java:
```diff
@@ -59,8 +62,12 @@ private void initLogWriter(Path logFilePath) {
-        ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
-        Runnable task = new Runnable() {
+        ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                .setNameFormat("event-logger-flush-%d")
```
should we add `setDaemon(true)` ?

satishd replied:
By default daemon is set as false, it should be set as true.

HeartSaVioR replied:
@arunmahadevan @satishd 

I intended to set daemon as false, since I would like to provide a chance to shutdown gracefully, via to wait for current running task to be finished instead of killing immediately.
Typically `shutdownNow()` will trigger interrupt and run() method looks like be able to handle interrupt (though it will throw RuntimeException).

If we are in favor of guaranteeing JVM shutdown more than graceful shutdown, I'm happy to  change it.

satishd replied:
I'm in favor of guaranteeing JVM shutdown in this case as I have seen dangling worker scenarios are hard for users to debug and maintain.

HeartSaVioR replied:
OK, makes sense. Will add `setDaemon(true)`.

HeartSaVioR replied:
Addressed.

arunmahadevan commented:
+1 LGTM, left one comment.

satishd commented on storm-client/src/jvm/org/apache/storm/metric/FileBasedEventLogger.java:
```diff
@@ -120,8 +127,30 @@ protected String buildLogMessage(EventInfo event) {
+        if (flushScheduler != null) {
+            flushScheduler.shutdown();
+            try {
+                if (!flushScheduler.awaitTermination(1, TimeUnit.SECONDS)) {
```
nit: You may want to have multiple iterations(which can be 2 or more) with while loop instead of nested if conditionals.

HeartSaVioR replied:
Most of cases in codebase we call shutdownNow() and doesn't wait more. I'll follow the case.

HeartSaVioR commented:
@arunmahadevan @satishd 
Thanks for reviewing. I addressed review comments. Please take a look again. I'll squash commits into one once I got +1. Thanks!

HeartSaVioR commented:
Thanks. Squashed.
